pkg/executor/delete.go

```diff
@@ -231,35 +249,41 @@ func (e *DeleteExec) deleteMultiTablesByChunk(ctx context.Context) error {
 			}
 		}
 	}
-
 	return e.removeRowsInTblRowMap(tblRowMap)
 }
 
 func (e *DeleteExec) removeRowsInTblRowMap(tblRowMap tableRowMapType) error {
 	for id, rowMap := range tblRowMap {
 		var err error
-		rowMap.Range(func(h kv.Handle, val []types.Datum) bool {
-			err = e.removeRow(e.Ctx(), e.tblID2Table[id], h, val)
+		rowMap.Range(func(h kv.Handle, val handleInfoPair) bool {
+			err = e.removeRow(e.Ctx(), e.tblID2Table[id], h, val.handleVal, val.posInfo)
 			return err == nil
 		})
 		if err != nil {
 			return err
 		}
 	}
 
+	e.Ctx().GetTableCtx().ResetExtraInfo()
+
 	return nil
 }
 
-func (e *DeleteExec) removeRow(ctx sessionctx.Context, t table.Table, h kv.Handle, data []types.Datum) error {
+func (e *DeleteExec) removeRow(ctx sessionctx.Context, t table.Table, h kv.Handle, data []types.Datum, posInfo *plannercore.TblColPosInfo) error {
+	e.Ctx().GetTableCtx().SetExtraIndexKeyPosInfo(posInfo.IndexesForDelete)
 	err := t.RemoveRecord(ctx.GetTableCtx(), h, data)
 	if err != nil {
+		// Don't try to use defer for performance.
+		e.Ctx().GetTableCtx().ResetExtraInfo()
 		return err
 	}
 	tid := t.Meta().ID
 	err = onRemoveRowForFK(ctx, data, e.fkChecks[tid], e.fkCascades[tid])
 	if err != nil {
+		e.Ctx().GetTableCtx().ResetExtraInfo()
 		return err
 	}
+	e.Ctx().GetTableCtx().ResetExtraInfo()
```

Comment:
It's easy to forget to reset in this way(so is L267, L277, L283), better to abstract a unified place which is guanrateed to be called, and do the cleanup or reset job there.